### PR TITLE
Backward compatibility for 4.3.1

### PR DESCRIPTION
**Pull request by calixtus**
This PR should allow people in the future to return to JabRef 4.3.1 after using 5.0, if they want to keep using an old version without support or development.
Those, who already tried the latest version, can use in windows "JabRef-downgrade-regpatch.reg".

Still got two questions:
1. How do we include this regpatch in the installer package and 
2. Who can test this. I don't have Java8 and JabRef 4.3.1 installed anymore...

fixes https://github.com/JabRef/jabref/issues/5933,
fixes https://github.com/JabRef/jabref/issues/6234,
refs https://github.com/JabRef/jabref/issues/4579

- [ ] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by koppor** [comment]
# Testing

First my results of testing...

## JabRef from this branch and then 4.3.1

1. Run JabRef from this branch using `./gradlew run`
2. Quit JabRef
3. Install corretto 4.3.1 jre

    choco install corretto8jre

4. Start JabRef 4.3.1

    "C:\Program Files\Corretto\jre8\bin\java.exe" -jar "JabRef-4.3.1.jar"

Output:

```
20:37:37.356 [AWT-EventQueue-0] INFO  org.jabref.gui.maintable.MainTable - Exception while setting column widths. Choosing default.
java.lang.NumberFormatException: For input string: "28.0"
        at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[?:1.8.0_252]
```

## JabRef 5.0, then this JabRef, then 4.3.1

1. Start JabRef 5.0
2. Reset JabRef preferences
3. Fix `JAVA_HOME`
4. Run JabRef from this branch using `./gradlew run`
5. Close JabRef
6. Start JabRef 4.3.1

    ```
    File: grouptree.fxml not found, attempting with camel case
    File: grouptree.css not found, attempting with camel case
    #
    # A fatal error has been detected by the Java Runtime Environment:
    #
    #  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x0000000066c30754, pid=11032, tid=0x0000000000002b50
    ```
7. Extract windows-amd64-1.8.0_172.tar.gz, use <https://stackoverflow.com/a/14480193./873282>. Rename `*.jar.jar` to `*.jar`
8. `C:\TEMP\jre-172\bin\java.exe -jar "JabRef-4.3.1.jar"`
9. Works fine
10. Close JabRef
11. `C:\TEMP\jre-172\bin\java.exe -jar "JabRef-3.7.jar"`
12. Works fine
13. Start JabRef 5.0
14. Rearrange table columns
15. Quit JabRef
16. Start JabRef 5.0
17. Table columns as configured
18. Quit JabRef
19. Start JabRef from this branch
20. Table columns are reset!
21. Reorder Table columns
22. Quit JabRef
22. Start JabRef
23. Table column order restored
22. Start JabRef 4.3.1
23. Strange column headings

   grafik

24. Preferences to not open

    21:06:05.611 [AWT-EventQueue-0] ERROR org.jabref.FallbackExceptionHandler - Uncaught exception occurred in Thread[AWT-EventQueue-0,6,main]
java.lang.NumberFormatException: For input string: "28.0"
        at java.lang.NumberFormatException.forInputString(Unknown Source) ~[?:1.8.0_172]

File: src/main/java/org/jabref/migrations/PreferencesMigrations.java [hunk 1]
```diff
@@ -308,6 +312,18 @@ static void upgradePreviewStyleAllowMarkdown(JabRefPreferences prefs) {
         prefs.setPreviewStyle(migratedStyle);
     }
 
+    static void changeColumnVariableNamesFor51(JabRefPreferences preferences) {
+        // The variable names have to be hardcoded, because they have changed between 5.0 and 5.1
```
Comment (koppor) [inline src/main/java/org/jabref/migrations/PreferencesMigrations.java:316]: For me, it is OK to have local string constants. Externalizing them feels over-engineered.

**Comment by calixtus** [comment]
Do the test results still apply after the phone call?

**Comment by Siedlerchr** [comment]
I think some of the failures with the preferences might be related to font face/font size. I remember that we changed that as well

**Comment by calixtus** [comment]
@koppor : Something more to test... 😁

**Comment by koppor** [comment]
TLDR: With the patch 7c260f5cc works fine.

# JabRef 4.3.1 -> JabRef 5.1-dev -> JabRef 4.3.1

- JVM crash with Coretto, therefore using -172
- `C:\TEMP\jre-172\bin\java.exe -jar "JabRef-4.3.1.jar" -d all`
- Reordered table columns and changed font size
- `gradlew run`
- ```text
  java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.String (java.lang.Integer and java.lang.String
  are in module java.base of loader 'bootstrap')
        at org.jabref.preferences.JabRefPreferences.get(JabRefPreferences.java:901) ~[main/:?]
        at org.jabref.migrations.PreferencesMigrations.restoreVariablesForBackwardCompatibility(PreferencesMigrations.java:431)
```
- Added a patch at 7c260f5cc.
- Starts
- Reordered table columns and changed font size
- `C:\TEMP\jre-172\bin\java.exe -jar "JabRef-4.3.1.jar" -d all`
- Starts fine (does have the font size configured for 4.3.1 not the one of 5.1-dev)
- Start 5.1
- Starts fine with large font size

# JabRef 5.1-dev -> JabRef 4.3.1 -> JabRef 5.1-dev

- `gradlew run --args="-d dall"`
  
    Error opening file '-d': Could not find a suitable import format.

- Manually reset preferences
- Close
- Start JabRef 4.3.1
- Works
- Start JabRef 5.0
- Works
- Change font size and table column order
- Start JabRef 5.1
- Close
- Start 4.3.1
- Starts fine

**Comment by koppor** [comment]
Good to go from my side. - Cannot add a test for 7c260f5, because mockito does not allow returning an `Integer` when a `String` is requested. I cannot mock `private final Preferences prefs;` inside the preferences. I had to mock that if I wanted to emulate the Integer vs. String thing (the expection happens in https://github.com/JabRef/jabref/blob/c7d77670154c1b09d8115013c8f3b4175906219e/src/main/java/org/jabref/preferences/JabRefPreferences.java#L901)

**Comment by calixtus** [comment]
Ouch, I ran the test but did not do `gradle run`. Thanks for fixing that.
